app/routes/auth.js

```diff
@@ -0,0 +1,180 @@
+            error = "Password is invalid"
+            break
+    }
+    res.render('pages/signup', { err: error })
+})
+
+// API route to sign a new user up.
+router.post('/signup', (req, res, next) => {
+    let salt = crypto.randomBytes(16)
+    let user
+    let attemptsLeft = 10000
+
+    let passwordSchema = new passwordValidator()
+        .is().min(8)
+        .is().max(100)
+        .has().not().spaces()
+        .has().digits(1)
+    
+    if (!passwordSchema.validate(req.body.password)) return res.redirect('/signup?error=invalid-password')
+    if (!emailValidator.validate(req.body.email)) return res.redirect('/signup?error=invalid-email')
+    if (req.body.fname) req.body.fname = req.body.fname.trim()
+    if (req.body.lname) req.body.lname = req.body.lname.trim()
+
+    crypto.pbkdf2(req.body.password, salt, 310000, 32, 'sha256', async function (err, hashedPassword) {
+        if (err) { res.redirect('/signup?error=other') }
+        while (true) {
+            try {
+                user = await prisma.user.create({
+                    data: {
+                        id: generateRandomUID(),
+                        email: req.body.email,
+                        hashed_password: hashedPassword,
+                        salt: salt,
+                        fname: req.body.fname,
+                        lname: req.body.lname
+                    }
+                })
+            } catch (err) {
+                console.log(err)
+                if (err instanceof PrismaClientKnownRequestError && err.code === "P2002" && err.meta.target[0] === 'id' && attemptsLeft > 0) {
+                    attemptsLeft--
+                    continue
+                } else {
+                    return res.redirect(`/signup?error=${attemptsLeft <= 0 ? 'other' : 'email'}`)
+                }
+            }
+            break
+        }
+
+        req.login(user, (err) => {
+            if (err) return next(err)
+            res.redirect('/')
+        })
+    })
+})
+
+// API route to log a user out.
+router.post('/logout', function (req, res, next) {
+    req.logout(function (err) {
+        if (err) { return next(err); }
+        let url = req.header('Referer') || '/'
+        res.redirect(url);
+    });
+});
+
+// Route for seeing user account data.
+router.get('/account', isLoggedIn, (req, res) => {
+    res.render('pages/account')
+})
+
+/**
+ * Generates a random user id where the first 3 chars are random lowercase alphabet letters and the last 4 chars are a zero-padded multiple of 37.
+ * @returns string containing the random user id
+ */
+function generateRandomUID() {
+    let userid = []
+    let upperLimit = 9999
+    let prefixLength = 3
+
+    for (let i = 0; i < prefixLength; i++) {
+        let n = Math.floor(Math.random() * 26)
+        userid.push(String.fromCharCode(97 + n))
+    }
+
+    let factor = Math.ceil(Math.random() * Math.floor(upperLimit / 37))
+    userid.push(String(factor * 37).padStart(4, '0'))
```

Comment:
Has a possibility of producing "0000" as postfix when `Math.random()` produces a 0 on line 164

Reply:
Fixed the bug by adding 1 (I think)n